src/algebra/non_unital_alg_hom.lean

```diff
@@ -0,0 +1,182 @@
+⟨to_distrib_mul_action_hom⟩
+
+instance : has_coe (non_unital_alg_hom R A B) (mul_hom A B) := ⟨to_mul_hom⟩
+
+@[simp] lemma to_distrib_mul_action_hom_eq_coe (f : non_unital_alg_hom R A B) :
+  f.to_distrib_mul_action_hom = ↑f :=
+rfl
+
+@[simp] lemma to_mul_hom_eq_coe (f : non_unital_alg_hom R A B) : f.to_mul_hom = ↑f :=
+rfl
+
+@[simp, norm_cast] lemma coe_to_distrib_mul_action_hom (f : non_unital_alg_hom R A B) :
+  ((f : A →+[R] B) : A → B) = f :=
+rfl
+
+@[simp, norm_cast] lemma coe_to_mul_hom (f : non_unital_alg_hom R A B) :
+  ((f : mul_hom A B) : A → B) = f :=
+rfl
+
+@[norm_cast] lemma coe_linear_map_mk (f : non_unital_alg_hom R A B) (h₁ h₂ h₃ h₄) :
+  ((⟨f, h₁, h₂, h₃, h₄⟩ : non_unital_alg_hom R A B) : A →+[R] B) =
+  ⟨f, h₁, h₂, h₃⟩ :=
+by { ext, refl, }
+
+@[norm_cast] lemma coe_mul_hom_mk (f : non_unital_alg_hom R A B) (h₁ h₂ h₃ h₄) :
+  ((⟨f, h₁, h₂, h₃, h₄⟩ : non_unital_alg_hom R A B) : mul_hom A B) = ⟨f, h₄⟩ :=
+by { ext, refl, }
+
+@[simp] lemma map_smul (f : non_unital_alg_hom R A B) (c : R) (x : A) :
+  f (c • x) = c • f x :=
+f.to_distrib_mul_action_hom.map_smul c x
+
+@[simp] lemma map_add (f : non_unital_alg_hom R A B) (x y : A) :
+  f (x + y) = (f x) + (f y) :=
+f.to_distrib_mul_action_hom.map_add x y
+
+@[simp] lemma map_mul (f : non_unital_alg_hom R A B) (x y : A) :
+  f (x * y) = (f x) * (f y) :=
+f.to_mul_hom.map_mul x y
+
+@[simp] lemma map_zero (f : non_unital_alg_hom R A B) : f 0 = 0 :=
+f.to_distrib_mul_action_hom.map_zero
+
+instance : has_zero (non_unital_alg_hom R A B) :=
+⟨{ map_mul' := by simp,
+   .. (0 : A →+[R] B) }⟩
+
+instance : has_one (non_unital_alg_hom R A A) :=
+⟨{ map_mul' := by simp,
+   .. (1 : A →+[R] A) }⟩
+
+@[simp] lemma coe_zero : ((0 : non_unital_alg_hom R A B) : A → B) = 0 := rfl
+
+@[simp] lemma coe_one : ((1 : non_unital_alg_hom R A A) : A → A) = id := rfl
+
+lemma zero_apply (a : A) : (0 : non_unital_alg_hom R A B) a = 0 := rfl
+
+lemma one_apply (a : A) : (1 : non_unital_alg_hom R A A) a = a := rfl
+
+instance : inhabited (non_unital_alg_hom R A B) := ⟨0⟩
+
+/-- The composition of morphisms is a morphism. -/
+def comp (f : non_unital_alg_hom R B C) (g : non_unital_alg_hom R A B) : non_unital_alg_hom R A C :=
+{ .. (f : mul_hom B C).comp (g : mul_hom A B),
+  .. (f : B →+[R] C).comp (g : A →+[R] B) }
+
+@[simp, norm_cast] lemma coe_comp (f : non_unital_alg_hom R B C) (g : non_unital_alg_hom R A B) :
+  (f.comp g : A → C) = (f : B → C) ∘ (g : A → B) :=
+rfl
+
+lemma comp_apply (f : non_unital_alg_hom R B C) (g : non_unital_alg_hom R A B) (x : A) :
+  f.comp g x = f (g x) :=
+rfl
+
+/-- The inverse of a bijective morphism is a morphism. -/
+def inverse (f : non_unital_alg_hom R A B) (g : B → A)
+  (h₁ : function.left_inverse g f) (h₂ : function.right_inverse g f) :
+  non_unital_alg_hom R B A :=
+{ .. (f : mul_hom A B).inverse g h₁ h₂,
+  .. (f : A →+[R] B).inverse g h₁ h₂ }
+
+@[simp] lemma coe_inverse (f : non_unital_alg_hom R A B) (g : B → A)
```

Comment:
I had forgotten I wrote this `simp` lemma here.

The alternative approach would be to use `simps` when defining `inverse` (and add `to_fun := g`) but I think I actually prefer this since it's a more general lemma.

Reply:
A possible downside of this simp lemma is that it applies before `map_add` has a chance to, a problem which the applied version doesn't share (I think)

Reply:
Good point though quick experiment indicates both possible lemmas suffer from this deficiency:
```lean
import algebra.non_unital_alg_hom

variables {R A B : Type*} [semiring R]
variables [non_unital_non_assoc_semiring A] [module R A]
variables [non_unital_non_assoc_semiring B] [module R B]
variables (f : non_unital_alg_hom R A B)
variables (g : B → A) (h₁ : function.left_inverse g f) (h₂ : function.right_inverse g f)

example (x y : B) : f.inverse g h₁ h₂ (x + y) = (g x) + (g y) :=
begin
  simp, -- applies only `non_unital_alg_hom.coe_inverse`
  -- Goal now: `g (x + y) = g x + g y`
  sorry,
end

-- Unnecessary as simp picks `inverse_apply` over `coe_inverse` below but maybe clarifies the point.
local attribute [-simp] non_unital_alg_hom.coe_inverse

@[simp] lemma non_unital_non_assoc_semiring.inverse_apply (x : B) :
  f.inverse g h₁ h₂ x = g x :=
rfl

example (x y : B) : f.inverse g h₁ h₂ (x + y) = (g x) + (g y) :=
begin
  simp, -- applies only `non_unital_non_assoc_semiring.inverse_apply`
  -- Goal now: `g (x + y) = g x + g y`
  sorry,
end
```